### PR TITLE
feat(tui): surface backpressure and dropped events (#130)

### LINKED ISSUE #130: TUI remote runtime: surface backpressure and dropped events
## Goal
Make event drops/backpressure visible instead of silent.

## Acceptance criteria
- Dropped events are counted and surfaced as a system/protocol event or transcript warning.
- The user can tell when output may be incomplete.
- Add tests for overflow/backpressure behavior.

### PR DESCRIPTION
**Pull request by lsm**
## Summary

Makes EventStream backpressure visible in the TUI instead of silently dropping events.

- `EventStream` now exposes `isFull()` so producers can detect the backpressure condition.
- `TuiRuntime` counts dropped events and emits a `system_warning` transcript event ("Warning: N events dropped due to backpressure").
- A `backpressure_status` event clears the active indicator once the ring buffer drains.
- `AppState` surfaces the warning and tracks `dropped_event_count` / `backpressure_active`.
- The status bar renders `backpressure:N` while active and `drops:N` after it clears.
- Added tests for ring-buffer overflow, warning emission, state application, and status-bar rendering.

## Test notes

- `zig build test-unit-core` passes.
- `zig build test-unit-tui` passes except for one pre-existing failure on `main` (`TuiModel remote streaming Enter falls back to submit and preserves composer`), which is unrelated to these changes.
- `./scripts/check-zig-patterns.sh` passes.

Closes #130

### CODE REVIEW

File: zig/src/tui/runtime.zig [hunk 1]
```diff
@@ -977,27 +929,68 @@ pub const TuiRuntime = struct {
...
+        const warning = TuiEvent{ .system_warning = .{ .message = OwnedSlice(u8).initOwned(message) } };
+        if (self.pushRaw(warning)) {
+            self.dropped_since_warning = 0;
+        }
+    }
+
+    fn clearBackpressureIfIdle(self: *TuiRuntime) void {
+        if (!self.backpressure_active) return;
+        if (self.event_stream.isFull()) return;
+        const status = TuiEvent{ .backpressure_status = .{ .active = false, .dropped_count = self.dropped_event_count } };
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/runtime.zig]: Added emitBackpressureActiveIfNeeded() in streamEvents(); it pushes backpressure_status{active=true} once the ring has a free slot, so AppState.backpressure_active is set while backpressure is ongoing. clearBackpressureIfIdle() later emits active=false when the queue drains.

File: zig/src/tui/runtime.zig [hunk 2]
```diff
@@ -977,27 +929,68 @@ pub const TuiRuntime = struct {
...
-                        var mutable = dropped;
-                        mutable.deinit(self.allocator);
-                    } else {
-                        std.Thread.yield() catch {};
-                    }
-                    continue;
-                },
-            };
+            self.flushDroppedWarning();
+            if (self.pushRaw(event)) {
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/runtime.zig]: Fixed pushTerminal to use pushUncounted() for the terminal event itself; only the evicted pending event is counted as a drop. pushCounted() is used for normal events and counts only actual drops.

File: zig/src/tui/runtime.zig [hunk 3]
```diff
@@ -977,27 +929,68 @@ pub const TuiRuntime = struct {
...
-    fn push(self: *TuiRuntime, event: TuiEvent) void {
-        self.event_stream.push(event) catch {
-            var mutable = event;
-            mutable.deinit(self.allocator);
+    fn pushRaw(self: *TuiRuntime, event: TuiEvent) bool {
+        self.event_stream.push(event) catch |err| switch (err) {
+            error.QueueFull => {
+                self.dropped_event_count += 1;
+                self.dropped_since_warning += 1;
+                self.backpressure_active = true;
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/runtime.zig]: Guarded dropped_event_count, dropped_since_warning, backpressure_active, and backpressure_status_active_emitted with backpressure_mutex. pushCounted, pushUncounted, flushDroppedWarning, clearBackpressureIfIdle, and emitBackpressureActiveIfNeeded all acquire the mutex before mutating state or pushing internal events.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Sonnet 4 (Anthropic)

> **Model:** claude-sonnet-4-20250514 | **Client:** NeoKai | **Provider:** Anthropic

**Recommendation: REQUEST_CHANGES**

### Summary

Overall this is a clean implementation. The backpressure detection, warning emission, status bar rendering, and test coverage are well-structured. Found one should-fix issue (P1) and a few suggestions.

### P1 — `stream_aborted` guard discards backpressure events

**File:** `zig/src/tui/state.zig:623-626`

When `stream_aborted` is true, only `.turn_end`, `.agent_end`, and `.@"error"` pass through `applyEvent`. The new `.backpressure_status` and `.system_warning` variants hit the `else => return` and are silently discarded.

**Impact:** If the user aborts a turn while backpressure is active, the status bar will permanently show `backpressure:N` because `self.backpressure_active` and `self.dropped_event_count` are never updated from the `backpressure_status { active = false }` event.

**Fix:** Add the new variants to the allowlist:

```zig
if (self.stream_aborted) switch (event) {
    .turn_end, .agent_end, .@"error", .backpressure_status, .system_warning => {},
    else => return,
};
```

### P2 — `backpressure_active` / `backpressure_status_active_emitted` read outside mutex in `streamEvents`

**File:** `zig/src/tui/runtime.zig:898`

```zig
if (self.backpressure_active and !self.event_stream.isFull()) {
    if (self.backpressure_status_active_emitted) {
```

Both `backpressure_active` and `backpressure_status_active_emitted` are plain `bool` fields protected by `backpressure_mutex` everywhere else (pushCounted, clearBackpressureIfIdle, etc.). This read path accesses them without the mutex. Technically a data race under Zig's memory model.

**Fix:** Either read under `backpressure_mutex` or make these `std.atomic.Value(bool)`.

### P2 — `pushTerminal` counter inflation under concurrent poll

**File:** `zig/src/tui/runtime.zig:1007-1017`

Counters are incremented *before* calling `self.event_stream.poll()`. If another consumer thread drains the stream between the failed push and the poll, `poll()` returns null and no event was actually evicted — but the counters are already incremented. The net effect is a slightly inflated drop count under concurrent consumers.

**Suggestion:** Move the counter increment after confirming `poll()` returned a non-null event. Alternatively, only count the eviction if `poll()` succeeds.

### P3 — Test coverage gaps

- The runtime test does not exercise `pushTerminal` (the eviction path).
- The runtime test does not verify `backpressure_status { active = true }` emission via `emitBackpressureActiveIfNeeded`.
- The `resetEventStreamForTurn` path is untested with stale backpressure state (P2 spurious `backpressure_status { active = false }` at start of new turn).

### Test results

- `test-unit-core`: ✅ all pass
- `test-unit-tui`: 345/346 pass (1 pre-existing failure on main: `TuiModel remote streaming Enter falls back to submit and preserves composer`)
- `check-zig-patterns.sh`: ✅ pass

File: zig/src/tui/runtime.zig [hunk 4]
```diff
@@ -890,6 +895,13 @@ pub const TuiRuntime = struct {
                 };
             }
         }
+        if (self.backpressure_active and !self.event_stream.isFull()) {
```
Comment (lsm) [inline zig/src/tui/runtime.zig]: **P2 — Data race: `backpressure_active` and `backpressure_status_active_emitted` are plain `bool` fields read here without `backpressure_mutex`.**

All writes to these fields are guarded by `backpressure_mutex` (in `pushCounted`, `clearBackpressureIfIdle`, `emitBackpressureActiveIfNeeded`). This read path is the exception. Technically a data race under Zig's memory model.

Fix: wrap in `backpressure_mutex` or convert to `std.atomic.Value(bool)`.

File: zig/src/tui/runtime.zig [hunk 5]
```diff
@@ -978,29 +990,113 @@ pub const TuiRuntime = struct {
...
-            };
-            return;
+            self.flushDroppedWarning();
+            if (self.pushUncounted(event)) {
+                self.clearBackpressureIfIdle();
+                return;
+            }
+            // The stream is still full. Evict the oldest pending event to make
+            // room for this terminal event, counting only the eviction as a drop.
+            while (!self.backpressure_mutex.tryLock()) std.atomic.spinLoopHint();
```
Comment (lsm) [inline zig/src/tui/runtime.zig:1080]: **P2 — Counter inflation under concurrent poll.**

Counters (`dropped_event_count`, `dropped_since_warning`) are incremented *before* `self.event_stream.poll()`. If another consumer thread drained the stream between the failed push and this poll, `poll()` returns null and no event was actually evicted — but the counters are already inflated by 1.

Suggestion: move counter increment after confirming `poll()` returned a non-null event, or use a local flag:

```zig
if (self.event_stream.poll()) |dropped| {
    while (!self.backpressure_mutex.tryLock()) std.atomic.spinLoopHint();
    self.dropped_event_count += 1;
    self.dropped_since_warning += 1;
    self.backpressure_active = true;
    self.backpressure_mutex.unlock();
    var mutable = dropped;
    mutable.deinit(self.allocator);
} else {
    std.Thread.yield() catch {};
}
```

File: zig/src/tui/state.zig [hunk 6]
```diff
@@ -682,6 +684,11 @@ pub const AppState = struct {
             },
             .context_usage => |payload| self.applyContextUsage(payload),
             .prompt_segment_usage => |payload| self.applyPromptSegmentUsage(payload),
```
Comment (lsm) [inline zig/src/tui/state.zig:851]: **P1 — `backpressure_status` and `system_warning` are silently discarded when `stream_aborted` is true (state.zig:623-626).**

The `stream_aborted` guard at line 623 only allows `.turn_end`, `.agent_end`, `.@"error"` through. The new `.backpressure_status` and `.system_warning` variants hit `else => return` and are discarded.

**Impact:** After abort while backpressure is active, status bar permanently shows stale `backpressure:N` because `self.backpressure_active` and `self.dropped_event_count` are never updated.

**Fix:** Add the new variants to the allowlist at line 624:

```zig
.turn_end, .agent_end, .@"error", .backpressure_status, .system_warning => {},
```

File: zig/src/tui/runtime.zig (hunk 4 above)
Comment (lsm) [reply zig/src/tui/runtime.zig]: Fixed: converted backpressure_active and backpressure_status_active_emitted to std.atomic.Value(bool). All reads now use .load(.acquire) and writes use .store(.release), so the UI thread in streamEvents() no longer races producer updates.

File: zig/src/tui/runtime.zig (hunk 5 above)
Comment (lsm) [reply zig/src/tui/runtime.zig:1080]: Fixed: in pushTerminal the counter increments now happen only after poll() confirms an event was actually evicted. If poll() returns null (another thread drained the stream), the counters are not inflated.

File: zig/src/tui/state.zig (hunk 6 above)
Comment (lsm) [reply zig/src/tui/state.zig:851]: Fixed: added .backpressure_status and .system_warning to the stream_aborted allowlist in state.zig, and added tests verifying the fields/transcript update while stream_aborted is true.

File: zig/src/tui/runtime.zig [hunk 7]
```diff
@@ -890,6 +895,13 @@ pub const TuiRuntime = struct {
                 };
             }
         }
+        if (self.backpressure_active and !self.event_stream.isFull()) {
+            if (self.backpressure_status_active_emitted) {
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/runtime.zig]: Fixed: the two status flags are now std.atomic.Value(bool), so streamEvents() reads them with .load(.acquire) before deciding to emit or clear, eliminating the race between UI and producer threads.

File: zig/src/tui/state.zig [hunk 8]
```diff
@@ -427,6 +427,8 @@ pub const AppState = struct {
     /// cancelled turn that are drained afterwards must not flip the status bar
     /// back to streaming.
     stream_aborted: bool = false,
+    dropped_event_count: u64 = 0,
+    backpressure_active: bool = false,
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/state.zig:534]: Fixed: resetReplayState() now clears dropped_event_count and backpressure_active, and a test verifies they are zeroed/false after reset.

File: zig/src/tui/runtime.zig [hunk 9]
```diff
@@ -978,29 +990,113 @@ pub const TuiRuntime = struct {
...
-                        mutable.deinit(self.allocator);
-                    } else {
-                        std.Thread.yield() catch {};
-                    }
-                    continue;
-                },
-            };
-            return;
+            self.flushDroppedWarning();
+            if (self.pushUncounted(event)) {
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/runtime.zig]: Fixed: pushTerminal now flushes any pending warning before entering the eviction loop, reserves freeSlots() >= 1 (or 2 when a warning is pending), then pushes the terminal event. flushDroppedWarning() is no longer called inside the eviction loop, so the warning cannot steal the slot reserved for the terminal event.

**Review by lsm — COMMENTED** [review]
@/tmp/review_body.md

File: zig/src/event_stream.zig [hunk 10]
```diff
@@ -351,6 +351,23 @@ pub fn EventStream(comptime T: type, comptime R: type) type {
...
+            const current_tail = self.tail.load(.acquire);
+            return ((current_head + 1) & RING_BUFFER_MASK) == current_tail;
+        }
+
+        pub fn freeSlots(self: *Self) usize {
+            self.mutex.lockUncancelable(defaultIo());
+            defer self.mutex.unlock(defaultIo());
+            const current_head = self.head.load(.acquire);
+            const current_tail = self.tail.load(.acquire);
+            const used = (current_head - current_tail) & RING_BUFFER_MASK;
```
Comment (lsm) [inline zig/src/event_stream.zig]: @/tmp/p0_comment.md

[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/event_stream.zig]: Fixed: freeSlots now uses wrapping subtraction (`-%`) before masking, so head < tail after a wrap no longer underflows. Added a wrap-around freeSlots test that cycles 400 events.

File: zig/src/tui/runtime.zig [hunk 11]
```diff
@@ -210,6 +210,11 @@ pub const TuiRuntime = struct {
     last_turn_stop_reason: ?ai_types.StopReason = null,
     compact_output: bool = false,
     run_async: bool = true,
+    dropped_event_count: u64 = 0,
+    dropped_since_warning: u64 = 0,
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/runtime.zig:214]: Fixed: resetEventStreamForTurn now clears dropped_event_count, dropped_since_warning, backpressure_active, and backpressure_status_active_emitted when the event stream is recreated for a new turn/session.

File: zig/src/tui/runtime.zig [hunk 12]
```diff
@@ -978,27 +990,124 @@ pub const TuiRuntime = struct {
...
+                self.backpressure_active.store(true, .release);
+                var mutable = dropped;
+                mutable.deinit(self.allocator);
+            } else {
+                std.Thread.yield() catch {};
+            }
+            self.backpressure_mutex.unlock();
+        }
+
+        if (warning_pending) {
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/runtime.zig]: Fixed: pushTerminal now calls flushDroppedWarning() after the terminal event is enqueued, so drops accumulated during the eviction loop are emitted even if warning_pending was false at loop entry.

File: zig/src/tui/app.zig [hunk 13]
```diff
@@ -796,8 +800,8 @@ pub const App = struct {
     }
 
     fn supportsStreamingShortcuts(self: *const App) bool {
-        _ = self.runtime orelse return self.session != null;
-        return self.session != null;
+        const runtime = self.runtime orelse return self.session != null;
+        return runtime.backend == .local;
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/app.zig]: Fixed: reverted supportsStreamingShortcuts to the pre-existing session-based gate so remote steering/follow-up queues are used during streaming. Updated the test to expect steer_count=1 (renamed it to reflect the actual behavior).

File: zig/src/tui/runtime.zig [hunk 14]
```diff
@@ -890,6 +895,13 @@ pub const TuiRuntime = struct {
                 };
             }
         }
+        if (self.backpressure_active.load(.acquire) and !self.event_stream.isFull()) {
+            if (self.backpressure_status_active_emitted.load(.acquire)) {
+                self.clearBackpressureIfIdle();
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/runtime.zig]: Fixed: switched backpressure status updates from event-driven to polling. streamEvents and push no longer emit backpressure_status events. App.drainEvents calls runtime.backpressureState() once per frame after draining, which reads the current state and auto-clears the active flag when the ring recovers. This guarantees the active indicator is rendered for at least one frame.

**Review by lsm — COMMENTED** [review]
@/tmp/approve_review.md

**Comment by lsm** [comment]
Merge conflict with `main` and 2 unresolved review threads (P2 recovery-path races from @chatgpt-codex-connector) blocking final approval.

**Action needed before merge:**
1. Rebase on `main` to resolve conflicts.
2. Resolve the 2 open review threads — either address the feedback or reply with reasoning and mark resolved.

Once conflicts resolved and threads addressed, approval can finalize. The P0 underflow fix and all round 3 P2 fixes are verified; remaining threads are recovery-path edge cases documented for follow-up.

File: zig/src/tui/runtime.zig [hunk 15]
```diff
@@ -978,27 +1010,97 @@ pub const TuiRuntime = struct {
...
+        // The warning is only pending if drops were counted after the flush above.
+        var warning_pending: bool = false;
+        while (!self.backpressure_mutex.tryLock()) std.atomic.spinLoopHint();
+        warning_pending = self.dropped_since_warning > 0;
+        self.backpressure_mutex.unlock();
+
+        const needed_free_slots: usize = if (warning_pending) 2 else 1;
+        while (self.event_stream.freeSlots() < needed_free_slots) {
+            while (!self.backpressure_mutex.tryLock()) std.atomic.spinLoopHint();
+            if (self.event_stream.poll()) |dropped| {
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/runtime.zig]: Fixed in 7a10824. pushTerminal no longer uses a freeSlots() capacity check; it attempts pushUncounted directly and only evicts when push returns QueueFull. This eliminates the race where the UI thread drains between a capacity check and a poll, preventing false evictions and inflated drop counts during recovery.

File: zig/src/tui/runtime.zig [hunk 16]
```diff
@@ -978,27 +1010,97 @@ pub const TuiRuntime = struct {
     }
 
     fn push(self: *TuiRuntime, event: TuiEvent) void {
-        self.event_stream.push(event) catch {
-            var mutable = event;
-            mutable.deinit(self.allocator);
-        };
+        self.flushDroppedWarning();
+        _ = self.pushCounted(event);
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/runtime.zig]: Fixed in 7a10824. push() now pushes the real event first via pushCounted, then calls flushDroppedWarning. A pending warning can no longer consume the only free slot and cause the event being pushed to be dropped. If the warning push fails (queue full after the event), it stays pending and is emitted on the next drain.

File: zig/src/tui/runtime.zig [hunk 17]
```diff
@@ -983,27 +1015,97 @@ pub const TuiRuntime = struct {
...
+        self.flushDroppedWarning();
+
+        // Determine whether we also need to reserve a slot for a pending warning.
+        // The warning is only pending if drops were counted after the flush above.
+        var warning_pending: bool = false;
+        while (!self.backpressure_mutex.tryLock()) std.atomic.spinLoopHint();
+        warning_pending = self.dropped_since_warning > 0;
+        self.backpressure_mutex.unlock();
+
+        const needed_free_slots: usize = if (warning_pending) 2 else 1;
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/runtime.zig]: Fixed in 7a10824. pushTerminal no longer pre-computes warning_pending or needed_free_slots. It flushes any pending warning once before the loop, then attempts pushUncounted directly; it only evicts when push returns QueueFull. New drops accumulated inside the loop have their warning delivered by the next push() pass (or the next flushDroppedWarning at the top of the next push/pushTerminal call), because a completed stream's final drops are inherently undeliverable until the next turn begins — the counters persist in backpressureState() so the status bar still reports them.

**Review by lsm — COMMENTED** [review]
@/tmp/final_review.md

File: zig/src/tui/app.zig [hunk 18]
```diff
@@ -789,6 +793,14 @@ pub const App = struct {
             try self.state.applyEvent(ev);
         }
         self.refreshQueuedCounts();
+        self.syncBackpressureState();
+    }
+
+    fn syncBackpressureState(self: *App) void {
+        const runtime = self.runtime orelse return;
+        const bp = runtime.backpressureState();
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/app.zig:1014]: Fixed in 47d6269. replaceMessages() now calls resetBackpressureState() for both remote and local session replacement paths, so Store.resumeSession()/replay cannot copy stale runtime dropped_event_count/backpressure_active back into AppState on the next sync.

File: zig/src/tui/runtime.zig [hunk 19]
```diff
@@ -898,6 +903,25 @@ pub const TuiRuntime = struct {
...
+    /// per frame after draining events so the status bar always reflects the
+    /// current runtime state without depending on event drain ordering.
+    pub fn backpressureState(self: *TuiRuntime) struct { active: bool, dropped_count: u64 } {
+        while (!self.backpressure_mutex.tryLock()) std.atomic.spinLoopHint();
+        defer self.backpressure_mutex.unlock();
+        // Auto-clear the internal flag when the ring has recovered so the
+        // status bar stops showing the active indicator without waiting for
+        // the next event push.
+        if (self.backpressure_active.load(.acquire) and !self.event_stream.isFull()) {
+            self.backpressure_active.store(false, .release);
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/runtime.zig]: Fixed in 47d6269. backpressureState() now captures the active/dropped snapshot before auto-clearing recovered queues, returns active=true for that recovery poll, then clears the internal flag so the next poll settles to drops:N. Updated the runtime regression test to verify the one-frame active report.

**Review by lsm — COMMENTED** [review]
@/tmp/final2_review.md

File: zig/src/tui/runtime.zig [hunk 20]
```diff
@@ -1014,33 +1054,71 @@ pub const TuiRuntime = struct {
...
+        self.pushDroppingOldestCounted(event);
+        self.flushDroppedWarning();
     }
 
-    fn pushDroppingOldest(self: *TuiRuntime, event: TuiEvent) void {
+    fn pushTerminal(self: *TuiRuntime, event: TuiEvent) void {
+        // Terminal events use the same preserve-newest path; they must not be
+        // dropped just because the projection queue is saturated.
+        self.pushDroppingOldestCounted(event);
+        self.flushDroppedWarning();
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/runtime.zig]: Fixed in 73f111d. pushTerminal now uses flushDroppedWarningDroppingOldest(), which forces terminal-path drop warnings into the queue by evicting oldest pending events if needed. Added regression coverage that fills the TUI queue, pushes a terminal event, and verifies both agent_end and system_warning are drained.

File: zig/src/tui/state.zig [hunk 21]
```diff
@@ -766,6 +770,11 @@ pub const AppState = struct {
             },
             .context_usage => |payload| self.applyContextUsage(payload),
             .prompt_segment_usage => |payload| self.applyPromptSegmentUsage(payload),
+            .system_warning => |payload| try self.appendTranscript(.system, payload.message.slice()),
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/state.zig]: Fixed in 73f111d. AppState now records system_warning as a visible warning/error transcript entry instead of .system, so chat mode renders the drop warning instead of folding it into hidden system-only conversation stats. Added transcript chat-mode regression coverage.

File: zig/src/tui/app.zig [hunk 22]
```diff
@@ -977,6 +981,7 @@ pub const App = struct {
             };
             self.refreshQueuedCounts();
         }
+        self.syncBackpressureState();
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/app.zig]: Fixed in 73f111d. App.drainEvents() now syncs runtime backpressure state immediately after draining and before auto-resuming queued work, so drops from the just-finished turn are captured in AppState before resumeSession() resets the runtime stream/counters.

**Review by lsm — COMMENTED** [review]
@/tmp/r6_review.md

File: zig/src/tui/app.zig [hunk 23]
```diff
@@ -975,6 +979,7 @@ pub const App = struct {
...
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/app.zig]: Fixed in e527423. drainEvents() now calls syncBackpressureState() immediately after refreshQueuedCounts() and before the auto-resume block, so drops from the completed turn are copied into AppState before resumeSession()/resetEventStreamForTurn() can clear runtime counters.

**Review by lsm — COMMENTED** [review]
@/tmp/r7_review.md

File: zig/src/tui/runtime.zig [hunk 24]
```diff
@@ -1014,33 +1054,108 @@ pub const TuiRuntime = struct {
...
-                    mutable.deinit(self.allocator);
-                    return;
-                },
-            };
-            return;
+            if (self.pushUncounted(event)) return;
+            // Stream is full. Evict the oldest pending event to make room for
+            // this event, counting only the confirmed eviction as a drop.
+            while (!self.backpressure_mutex.tryLock()) std.atomic.spinLoopHint();
+            if (self.event_stream.poll()) |dropped| {
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/runtime.zig:1095]: Addressed in 0e6d312: pushDroppingOldestCounted now retries event_stream.push() after taking backpressure_mutex and before evicting. If the UI consumer freed a slot after the initial QueueFull, the event enqueues without counting or dropping an older pending event. Only a second QueueFull under the lock proceeds to confirmed eviction/drop accounting.
